pants-plugins/examples/bash/shellcheck.py

```diff
@@ -0,0 +1,192 @@
+    field_set_type = ShellcheckFieldSet
+
+
+@rule
+async def run_shellcheck(
+    request: ShellcheckRequest, shellcheck: Shellcheck
+) -> LintResults:
+    if shellcheck.options.skip:
+        return LintResults()
+
+    # Shellcheck looks at direct dependencies to make sure that every symbol is defined, so we must
+    # include those in the run.
+    all_dependencies = await MultiGet(
+        Get(Targets, DependenciesRequest(field_set.dependencies))
+        for field_set in request.field_sets
+    )
+    # Now that we have all dependencies, we flatten the results into a single list of `BashSources`
+    # fields and we filter out all targets without a `BashSources` field registered because those
+    # dependencies are irrelevant to Shellcheck.
+    dependencies_sources_fields = [
+        tgt[BashSources]
+        for dependencies in all_dependencies
+        for tgt in dependencies
+        if tgt.has_field(BashSources)
+    ]
+
+    sources_request = Get(
+        SourceFiles,
+        AllSourceFilesRequest(
+            [
+                *(field_set.sources for field_set in request.field_sets),
+                *dependencies_sources_fields,
+            ]
+        ),
+    )
+
+    download_shellcheck_request = Get(
+        DownloadedExternalTool,
+        ExternalToolRequest,
+        shellcheck.get_request(Platform.current),
+    )
+
+    # If the user specified `--shellcheck-config`, we must search for the file they specified with
+    # `PathGlobs` to include it in the `input_digest`. We error if the file cannot be found.
+    config_snapshot_request = Get(
+        Snapshot,
+        PathGlobs(
+            globs=[shellcheck.options.config] if shellcheck.options.config else [],
+            glob_match_error_behavior=GlobMatchErrorBehavior.error,
+            description_of_origin="the option `--shellcheck-config`",
+        ),
+    )
+
+    sources, downloaded_shellcheck, config_snapshot = await MultiGet(
+        sources_request, download_shellcheck_request, config_snapshot_request
+    )
+
+    # The Process needs one single `Digest`, so we merge everything together.
+    input_digest = await Get(
+        Digest,
+        MergeDigests(
+            (
+                sources.snapshot.digest,
+                downloaded_shellcheck.digest,
+                config_snapshot.digest,
+            )
+        ),
+    )
+
+    process_result = await Get(
+        FallibleProcessResult,
+        Process(
+            argv=[
+                downloaded_shellcheck.exe,
+                *shellcheck.options.args,
+                *sources.snapshot.files,
+            ],
+            input_digest=input_digest,
+            description=f"Run Shellcheck on {pluralize(len(request.field_sets), 'target')}.",
+        ),
+    )
+    result = LintResult.from_fallible_process_result(
+        process_result, linter_name="Shellcheck"
+    )
+    return LintResults([result])
+
+
+def rules():
+    return [
+        run_shellcheck,
```

Comment:
Comments for each of these would be great too.

Reply:
I don't think it would be particularly helpful - every file is going to have this type of setup, and we document it on https://www.pantsbuild.org/v2.0/docs/plugins-new-linter pretty extensively.

Most of the comments in this file do not cover things mentioned on that guide. For example, we don't explain `GlobMatchErrorBehavior.error` on that guide.